common/tools/dev-tool/src/commands/samples/checkNodeVers.ts

```diff
@@ -0,0 +1,323 @@
+    buildRunSamplesScript(containerWorkspacePath, packageName, envFileName, logFilePath),
+    { mode: S_IRWXO }
+  );
+  callback?.();
+}
+
+function runDockerContainer(
+  dockerContextDirectory: string,
+  dockerImageName: string,
+  dockerContainerName: string,
+  containerWorkspace: string,
+  stdoutListener: (chunk: string | Buffer) => void,
+  stderrListener: (chunk: string | Buffer) => void,
+  callback?: () => void
+) {
+  const args = [
+    "run",
+    "--name",
+    dockerContainerName,
+    "--workdir",
+    containerWorkspace,
+    "-v",
+    `${dockerContextDirectory}:${containerWorkspace}`,
+    dockerImageName,
+    "./run_samples.sh"
+  ];
+  const dockerContainerRunProcess = pr.spawn("docker", args, {
+    cwd: dockerContextDirectory
+  });
+  log.info(`Started running the docker container ${dockerContainerName}`);
+  dockerContainerRunProcess.stdout.on("data", stdoutListener);
+  dockerContainerRunProcess.stderr.on("data", stderrListener);
+  dockerContainerRunProcess.on("exit", function(code) {
+    if (code === 0) {
+      log.info("Docker container finished running");
+      callback?.();
+    } else {
+      log.error("Docker container encountered an error");
+    }
+  });
+}
+
+export const commandInfo = makeCommandInfo(
+  "checkNodesVers",
+  "execute samples with different node versions, typically in preparation for release",
+  {
+    package_path: {
+      kind: "string",
+      description: "Path to the downloaded package built by the release pipeline"
+    },
+    directory: {
+      kind: "string",
+      description: "Base dir, default is process.cwd()",
+      default: process.cwd()
+    },
+    node_versions: {
+      kind: "string",
+      description: "A comma separated list of node versions to use",
+      default: "8,10,12"
+    },
+    context_directory_path: {
+      kind: "string",
+      description: "Absolute path to a directory used for mounting inside docker containers",
+      default: ""
+    },
+    keep_docker_context: {
+      kind: "boolean",
+      description: "Boolean to indicate whether to keep the current docker context directory",
+      default: false
+    },
+    log_in_file: {
+      kind: "boolean",
+      description:
+        "Boolean to indicate whether to save the the stdout and sterr for npm commands to the log.txt log file",
+      default: true
+    },
+    use_existing_docker_container: {
+      kind: "boolean",
+      description: "Boolean to indicate whether to use existing docker containers if any",
+      default: false
+    },
+    keep_docker_container: {
+      kind: "boolean",
+      description: "Boolean to indicate whether to keep docker containers",
+      default: false
+    },
+    keep_docker_image: {
+      kind: "boolean",
+      description: "Boolean to indicate whether to keep the downloaded docker images",
+      default: false
+    }
+  }
+);
+
+export default leafCommand(commandInfo, async (options) => {
```

Comment:
I find it hard to follow the callback logic. Can these be made async?

Reply:
Done.